wazp/utils.py

```diff
@@ -142,110 +145,340 @@ def export_selected_rows_as_yaml(
+    ]
+
+    return Polygon(list_points)
+
+
+def add_ROIs_to_video_dataframe(
+    df: pd.DataFrame, metadata: dict, ROIs_as_polygons: dict, app_storage: dict
+) -> pd.DataFrame:
+    """Assign ROI to every row in the dataframe
+    corresponding to the current video.
+
+    If no ROIs are defined for this video, an empty string
+    is assigned to all rows.
+
+    The hierarchy of ROIs is defined either based on their area
+    (if a point is contained in several ROIs, the smallest one
+    is assigned), or based on the custom order defined by the user.
+
+    Parameters
+    ----------
+    df : pd.Dataframe
+        pandas dataframe with the pose estimation results
+        for one video. It is a single-level dataframe, restructured
+        from the DeepLabCut output.
+    metadata : dict
+        dictionary with the metadata for the current video.
+    ROIs_as_polygons : dict
+        dictionary for the current video with ROI tags as keys,
+        and their corresponding shapely polygons as values.
+    app_storage : dict
+        data held in temporary memory storage,
+        accessible to all tabs in the app
+
+    Returns
+    -------
+    df : pd.Dataframe
+        pandas dataframe with the pose estimation results
+        for one video, and the ROI per bodypart per frame
+        assigned.
+    """
+
+    # Define hierarchy of ROIs
+    flag_use_ROI_custom_order = app_storage["config"].get(
+        "use_ROIs_order_as_hierarchy", False  # reads a bool
+    )
+
+    # -----------------
+    # Use custom hierarchy of ROIs
+    # if hierarchy flag in input config exists
+    # and is True
+    if flag_use_ROI_custom_order:
+        list_sorted_ROI_pairs = [
+            x
+            for x, _ in sorted(
+                zip(
+                    ROIs_as_polygons.items(),
+                    enumerate(app_storage["config"]["ROI_tags"]),
+                ),
+                key=lambda pair: pair[1][0],
+            )
+        ]
+
+    # else: set hierarchy of ROIs based in their area
+    else:
+        list_sorted_ROI_pairs = sorted(
+            ROIs_as_polygons.items(),
+            key=lambda pair: pair[1].area,  # sort by increasing area
         )
 
-        # select subset of rows based on
-        # frame numbers from slider (both inclusive)
-        # (index as frame number)
-        list_df_to_export.append(
-            df[
-                (df.index >= frame_start_end[0])
-                & (df.index <= frame_start_end[1])
-            ]
+    # -------------------
+    # Assign ROIs
+    for ROI_str, ROI_poly in list_sorted_ROI_pairs:
+        # for optimized performance
+        # (applies transform in place)
+        shapely.prepare(ROI_poly)
+
+        # Consider buffer around boundaries if required
+        # TODO: remove this buffer option? inspired by this SO answer
+        # https://stackoverflow.com/a/59033011
+        if "buffer_around_ROIs_boundaries" in app_storage["config"]:
```

Comment:
I think it could be useful for the user to define a 'thickness' for the region of interest. If it is not accessible by the user, then maybe we shouldn't have it at all (a buffer of 1e-9 seems as arbitrary as any other?).

Will add this to the list of 'things to ask user feedback on'

Reply:
I think it may be useful for the user to define a thickness for the ROIs, but maybe this is a point to get user feedback on. I don't think the user would have a problem understanding its meaning or setting its value. If we don't allow the user to edit it, then I would argue we don't need it at all (1e-09 seems as arbitrary as any other value?). 

Note that for now this parameter is not mandatory, so if it is not specified in the config then no buffer/thickness is set. We can also set its default value to 1e-09 instead.

For now I will leave it as it is (as an optional parameter) since that seems the most flexible, but I started an issue for things to get user feedback on (#69) and added it to the list. Let me know if you think otherwise.